client: Allow setting zap logger in config

serathius opened the pull request:
Support direct passing logger to clientv3. 
Ref https://github.com/etcd-io/etcd/issues/12326

clientv3 technically allows to achieve this by using `c := c.WIthLogger(logger)`, but this is error prone as some objects take logger before wrapping. 

cc @ptabor

gyuho commented:
lgtm, defer to @ptabor

ptabor reviewed (approved):
1. Please mark: 
  func (c *Client) WithLogger(lg *zap.Logger) *Client {

as deprecated and point on Config.Logger as canonical way of configuring client's logger.

2. Please change 1-2 tests invocation to have test-coverage of the new way. 
For example: 
  - client/v3/client_test.go:31
  - tests/integration/cluster_direct.go:48

serathius commented:
If we are deprecating `WithLogger` then we should also deprecate/update client constructors that depend on `WithLogger` as a only way to configure logging. Example https://github.com/etcd-io/etcd/blob/e0a8484c8f07422785e5826c576580973aeef4bc/client/v3/client.go#L83-L89 What do you think?

serathius commented:
Added usage of Logger in tests and replaced almost all usage of `WithLogger` with new field. Only one left is in https://github.com/etcd-io/etcd/blob/e0a8484c8f07422785e5826c576580973aeef4bc/server/etcdserver/api/v3client/v3client.go#L31-L35 as it uses `NewCtxClient` mentioned above.

ptabor commented:
> If we are deprecating `WithLogger` then we should also deprecate/update client constructors that depend on `WithLogger` as a only way to configure logging. Example
> 
> https://github.com/etcd-io/etcd/blob/e0a8484c8f07422785e5826c576580973aeef4bc/client/v3/client.go#L83-L89
> 
> What do you think?

WithLogger should be constructor's option ideally of this constructors. 
Currently this is executed post-construction and overrides loggers in the main object, but not loggers copied to all helper objects (like lease manager) that makes it very subtle and error prone.

But let's consider this for follow up PR.

ptabor commented:
Thank you for all the fixes.
